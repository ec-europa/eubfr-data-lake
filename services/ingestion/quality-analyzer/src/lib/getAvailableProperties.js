// @flow

const has = (object, key) =>
  object ? hasOwnProperty.call(object, key) : false;

const specialFields = [
  // monetary
  'total_cost',
  'eu_contrib',
  'private_fund',
  'public_fund',
  'other_contrib',
  // we only need to know if centroid is present
  'centroid',
  // we care only for the coordinates
  'location',
];

/**
 * Takes a object with exceptional structure and returns part of the initial object which is actually useful.
 *
 * @param {Object} initialObject The object representing a special value of a field.
 * @returns {Object} Only a selected part of the {initialObject} which is actually needed.
 *
 * @example
 * input => { "value": 20511539, "currency": "EUR", "raw": "20511539" }
 * output => { "value": 20511539 }
 *
 * @example
 * We care only about the value of a monetary field, new normalized structure is returned.
 *
 * input => { "value": 20511539, "currency": "EUR", "raw": "20511539" }
 * output => { "value": 20511539 }
 *
 * We care only if a centroid has latitude or longitude, `lat` and `lon` fields are skipped.
 *
 * input => { "lat": 52.022163, "lon": 20.914336 }
 * output => { skip: true }
 *
 * We care only if coordinates are present, `coordinates` field is preseved.
 *
 * input => { "type": "Point", "coordinates": [20.914336, 52.022163] }
 * output => { coordinates: true }
 *
 */
const getNormalizedObject = (initialObject: Object): Object => {
  // A monetary field in budget, value is enough of the all
  if (has(initialObject, 'currency')) {
    return { value: initialObject.value };
  }

  // A centroid with latitude and longitude, skip as we care only that it has .centroid
  // This is how we IGNORE the property/field of current level
  if (has(initialObject, 'lat') && has(initialObject, 'lon')) {
    return { skip: true };
  }

  // We care only if there are coordinates or not, not about their values
  // This is how we KEEP the property/field of current level
  if (has(initialObject, 'type') && has(initialObject, 'coordinates')) {
    return { coordinates: true };
  }

  return initialObject;
};

/**
 * Flags whether a given value is useful or not.
 *
 * @param {any} input The input to varify
 *
 * @returns {boolean} Whether acceptable.
 *
 */
const hasUsefulData = input => {
  const unAcceptableValues = [0, '0', 'EUR', '', null];

  if (!unAcceptableValues.includes(input)) {
    return true;
  }

  return false;
};

/**
 * Get information about which fields of an object are not empty.
 * To use, initiate `results` out of the scope of this function and call it on it.
 * Function does not return but accumulates results in `results`.
 *
 * @param {Object} obj Input object to inspect.
 * @param {Array<string>} results The accumulator of results after function's invocation.
 * @param {string} stack The string representation of a nested field in an object as deep as necessary.
 *
 * @returns {boolean} Whether acceptable.
 *
 */
const getAvailableProperties = (
  obj: Object,
  results: Array<string>,
  stack: string = ''
): void => {
  if (!obj) {
    return;
  }

  Object.keys(obj).forEach(property => {
    if (typeof obj[property] === 'object') {
      const fieldIsSpecial = specialFields.includes(property);

      const ObjectUnderInvestigation = fieldIsSpecial
        ? getNormalizedObject(obj[property])
        : obj[property];

      // If no previous stack, then property is sufficient as a root.
      const setStack = stack === '' ? property : `${stack}.${property}`;
      // Re-iterate
      getAvailableProperties(ObjectUnderInvestigation, results, setStack);
    } else if (hasUsefulData(obj[property]) && !has(results, stack)) {
      // nothing nested, make this property the root of the stack for this property
      if (stack === '') {
        results.push(property);
      } else {
        // The stack is the nested fields information
        const setStack = property === 'skip' ? stack : `${stack}.${property}`;
        results.push(setStack);
      }
    }
  });
};

export default getAvailableProperties;
