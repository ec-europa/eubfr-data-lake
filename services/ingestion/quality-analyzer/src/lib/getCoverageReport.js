// @flow

/**
 * Counts fields occurences.
 *
 * @param {Object} analyzedObject
 * @param {Object} results The outcome of the analysis
 *
 * @returns {Object} The results for the report.
 *
 */
const counter = (analyzedObject: Object): Object => {
  const results = {};

  analyzedObject.forEach(field => {
    // existing field
    if (results[field]) {
      results[field] += 1;
      // new field
    } else {
      results[field] = 1;
    }
  });

  return results;
};

/**
 * Flags whether a given value is useful or not.
 *
 * @param {Object} input The object for analysis.
 * @param {number} numRecords The number of overall occurences known outside of the scope of this function.
 *
 * @returns {Object} The results for the report.
 *
 */
const getCoverageReport = (
  input: Object,
  numRecords: number
): Array<Object> => {
  if (!input) {
    return [{}];
  }

  const results = counter(input);

  return Object.keys(results).map(field => {
    // Divide the number of occurences by the overall number of records
    const fieldCoverage = (results[field] / numRecords * 100).toFixed(2);

    return {
      [field]: results[field],
      coverage: fieldCoverage,
    };
  });
};

export default getCoverageReport;
