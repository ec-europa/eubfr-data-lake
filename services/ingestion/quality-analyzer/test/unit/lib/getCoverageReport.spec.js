/**
 * @jest-environment node
 */

import fs from 'fs';
import path from 'path';
import split2 from 'split2';
import getAvailableProperties from '../../../src/lib/getAvailableProperties';
import getCoverageReport from '../../../src/lib/getCoverageReport';

describe(`Helper getCoverageReport "@eubfr/ingestion-quality-analyzer" provides reports for number of occurences of a given field and its coverage`, () => {
  const transforms = [
    'agri/csv',
    'budg/xls',
    'iati/csv',
    'inforegio/json',
    'inforegio/xml',
    'valor/xls',
    'wifi4eu/xls',
  ];

  transforms.forEach(transform => {
    test(`Works with results of ${transform}`, () => {
      let numRecords = 0;
      const results = [];

      // In actual implementation this comes from s3 file meta.
      // Preserved in order to keep the structure of the results as close as possible to actual implementation.
      const fileMeta = {
        computed_key: 'foo',
        created_by: 'bar',
        last_modified: 'baz',
      };

      const file = path.resolve(
        `${__dirname}/../../stubs/${transform}/harmonized.ndjson`
      );

      const fileReadStream = fs.createReadStream(file);

      return new Promise((resolve, reject) => {
        fileReadStream
          .on('error', e => reject(e))
          .pipe(split2(JSON.parse))
          .on('data', record => {
            numRecords += 1;
            getAvailableProperties(record, results);
          })
          .on('finish', () => {
            const fieldStats = getCoverageReport(results, numRecords);
            const report = { meta: fileMeta, report: fieldStats };
            expect(report).toMatchSnapshot();
            resolve();
          });
      });
    });
  });
});
