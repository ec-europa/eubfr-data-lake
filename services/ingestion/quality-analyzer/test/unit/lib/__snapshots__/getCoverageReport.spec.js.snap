// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper getCoverageReport "@eubfr/ingestion-quality-analyzer" provides reports for number of occurences of a given field and its coverage Works with results of agri/csv 1`] = `
Object {
  "meta": Object {
    "computed_key": "foo",
    "created_by": "bar",
    "last_modified": "baz",
  },
  "report": Array [
    Object {
      "budget.eu_contrib": 6,
      "coverage": "200.00",
    },
    Object {
      "budget.funding_area": 3,
      "coverage": "100.00",
    },
    Object {
      "budget": 3,
      "coverage": "100.00",
    },
    Object {
      "coverage": "100.00",
      "description": 3,
    },
    Object {
      "coverage": "100.00",
      "ec_priorities": 3,
    },
    Object {
      "coverage": "200.00",
      "media.0": 6,
    },
    Object {
      "coverage": "100.00",
      "media.0.meta": 3,
    },
    Object {
      "coverage": "100.00",
      "programme_name": 3,
    },
    Object {
      "coverage": "100.00",
      "project_id": 3,
    },
    Object {
      "coverage": "100.00",
      "project_locations.0": 3,
    },
    Object {
      "coverage": "200.00",
      "project_locations.0.centroid": 6,
    },
    Object {
      "coverage": "100.00",
      "project_locations.0.location": 3,
    },
    Object {
      "coverage": "200.00",
      "project_locations.0.location.coordinates": 6,
    },
    Object {
      "coverage": "100.00",
      "results": 3,
    },
    Object {
      "coverage": "200.00",
      "third_parties.0": 6,
    },
    Object {
      "coverage": "200.00",
      "timeframe": 6,
    },
    Object {
      "coverage": "100.00",
      "title": 3,
    },
    Object {
      "budget.total_cost": 4,
      "coverage": "133.33",
    },
    Object {
      "coverage": "66.67",
      "project_website": 2,
    },
  ],
}
`;

exports[`Helper getCoverageReport "@eubfr/ingestion-quality-analyzer" provides reports for number of occurences of a given field and its coverage Works with results of budg/xls 1`] = `
Object {
  "meta": Object {
    "computed_key": "foo",
    "created_by": "bar",
    "last_modified": "baz",
  },
  "report": Array [
    Object {
      "action": 3,
      "coverage": "100.00",
    },
    Object {
      "budget.eu_contrib": 6,
      "coverage": "200.00",
    },
    Object {
      "call_year": 3,
      "coverage": "100.00",
    },
    Object {
      "coverage": "100.00",
      "description": 3,
    },
    Object {
      "coverage": "100.00",
      "programme_name": 3,
    },
    Object {
      "coverage": "100.00",
      "project_id": 3,
    },
    Object {
      "coverage": "100.00",
      "project_locations.0": 3,
    },
    Object {
      "coverage": "100.00",
      "project_locations.1": 3,
    },
    Object {
      "coverage": "100.00",
      "project_locations.2": 3,
    },
    Object {
      "coverage": "100.00",
      "project_locations.3": 3,
    },
    Object {
      "coverage": "66.67",
      "project_locations.4": 2,
    },
    Object {
      "coverage": "66.67",
      "project_locations.5": 2,
    },
    Object {
      "coverage": "200.00",
      "results": 6,
    },
    Object {
      "coverage": "100.00",
      "status": 3,
    },
    Object {
      "coverage": "100.00",
      "sub_programme_name": 3,
    },
    Object {
      "coverage": "100.00",
      "success_story": 3,
    },
    Object {
      "coverage": "666.67",
      "third_parties.0": 20,
    },
    Object {
      "coverage": "700.00",
      "third_parties.1": 21,
    },
    Object {
      "coverage": "700.00",
      "third_parties.2": 21,
    },
    Object {
      "coverage": "666.67",
      "third_parties.3": 20,
    },
    Object {
      "coverage": "700.00",
      "third_parties.4": 21,
    },
    Object {
      "coverage": "466.67",
      "third_parties.5": 14,
    },
    Object {
      "coverage": "200.00",
      "timeframe": 6,
    },
    Object {
      "coverage": "100.00",
      "title": 3,
    },
    Object {
      "coverage": "33.33",
      "project_locations.6": 1,
    },
    Object {
      "coverage": "33.33",
      "project_locations.7": 1,
    },
    Object {
      "coverage": "33.33",
      "project_locations.8": 1,
    },
    Object {
      "coverage": "33.33",
      "project_locations.9": 1,
    },
    Object {
      "coverage": "33.33",
      "project_locations.10": 1,
    },
    Object {
      "coverage": "33.33",
      "project_locations.11": 1,
    },
    Object {
      "coverage": "33.33",
      "project_locations.12": 1,
    },
    Object {
      "coverage": "33.33",
      "project_locations.13": 1,
    },
    Object {
      "coverage": "33.33",
      "project_locations.14": 1,
    },
    Object {
      "coverage": "233.33",
      "third_parties.6": 7,
    },
    Object {
      "coverage": "233.33",
      "third_parties.7": 7,
    },
    Object {
      "coverage": "233.33",
      "third_parties.8": 7,
    },
    Object {
      "coverage": "233.33",
      "third_parties.9": 7,
    },
    Object {
      "coverage": "233.33",
      "third_parties.10": 7,
    },
    Object {
      "coverage": "233.33",
      "third_parties.11": 7,
    },
    Object {
      "coverage": "233.33",
      "third_parties.12": 7,
    },
    Object {
      "coverage": "233.33",
      "third_parties.13": 7,
    },
    Object {
      "coverage": "233.33",
      "third_parties.14": 7,
    },
    Object {
      "coverage": "233.33",
      "third_parties.15": 7,
    },
    Object {
      "coverage": "233.33",
      "third_parties.16": 7,
    },
  ],
}
`;

exports[`Helper getCoverageReport "@eubfr/ingestion-quality-analyzer" provides reports for number of occurences of a given field and its coverage Works with results of iati/csv 1`] = `
Object {
  "meta": Object {
    "computed_key": "foo",
    "created_by": "bar",
    "last_modified": "baz",
  },
  "report": Array [
    Object {
      "coverage": "100.00",
      "description": 3,
    },
    Object {
      "coverage": "100.00",
      "project_id": 3,
    },
    Object {
      "coverage": "100.00",
      "project_locations.0": 3,
    },
    Object {
      "coverage": "100.00",
      "reporting_organisation": 3,
    },
    Object {
      "coverage": "166.67",
      "timeframe": 5,
    },
    Object {
      "coverage": "100.00",
      "title": 3,
    },
    Object {
      "budget.eu_contrib": 2,
      "coverage": "66.67",
    },
  ],
}
`;

exports[`Helper getCoverageReport "@eubfr/ingestion-quality-analyzer" provides reports for number of occurences of a given field and its coverage Works with results of inforegio/json 1`] = `
Object {
  "meta": Object {
    "computed_key": "foo",
    "created_by": "bar",
    "last_modified": "baz",
  },
  "report": Array [
    Object {
      "budget.total_cost": 6,
      "coverage": "200.00",
    },
    Object {
      "budget.eu_contrib": 6,
      "coverage": "200.00",
    },
    Object {
      "budget.funding_area": 2,
      "coverage": "66.67",
    },
    Object {
      "coverage": "100.00",
      "description": 3,
    },
    Object {
      "coverage": "100.00",
      "period": 3,
    },
    Object {
      "coverage": "100.00",
      "project_id": 3,
    },
    Object {
      "coverage": "200.00",
      "project_locations.0": 6,
    },
    Object {
      "coverage": "200.00",
      "project_locations.0.nuts.0": 6,
    },
    Object {
      "coverage": "100.00",
      "project_website": 3,
    },
    Object {
      "coverage": "266.67",
      "themes": 8,
    },
    Object {
      "coverage": "400.00",
      "third_parties.0": 12,
    },
    Object {
      "coverage": "66.67",
      "timeframe": 2,
    },
    Object {
      "coverage": "100.00",
      "title": 3,
    },
    Object {
      "coverage": "100.00",
      "type": 3,
    },
  ],
}
`;

exports[`Helper getCoverageReport "@eubfr/ingestion-quality-analyzer" provides reports for number of occurences of a given field and its coverage Works with results of inforegio/xml 1`] = `
Object {
  "meta": Object {
    "computed_key": "foo",
    "created_by": "bar",
    "last_modified": "baz",
  },
  "report": Array [
    Object {
      "budget.total_cost": 6,
      "coverage": "200.00",
    },
    Object {
      "budget.eu_contrib": 6,
      "coverage": "200.00",
    },
    Object {
      "coverage": "100.00",
      "description": 3,
    },
    Object {
      "coverage": "100.00",
      "period": 3,
    },
    Object {
      "coverage": "100.00",
      "project_id": 3,
    },
    Object {
      "coverage": "200.00",
      "project_locations.0": 6,
    },
    Object {
      "coverage": "100.00",
      "project_locations.0.nuts.0": 3,
    },
    Object {
      "coverage": "100.00",
      "project_website": 3,
    },
    Object {
      "coverage": "100.00",
      "themes": 3,
    },
    Object {
      "coverage": "100.00",
      "title": 3,
    },
    Object {
      "coverage": "100.00",
      "type": 3,
    },
  ],
}
`;

exports[`Helper getCoverageReport "@eubfr/ingestion-quality-analyzer" provides reports for number of occurences of a given field and its coverage Works with results of valor/xls 1`] = `
Object {
  "meta": Object {
    "computed_key": "foo",
    "created_by": "bar",
    "last_modified": "baz",
  },
  "report": Array [
    Object {
      "budget.eu_contrib": 6,
      "coverage": "200.00",
    },
    Object {
      "call_year": 3,
      "coverage": "100.00",
    },
    Object {
      "coverage": "100.00",
      "description": 3,
    },
    Object {
      "coverage": "100.00",
      "programme_name": 3,
    },
    Object {
      "coverage": "100.00",
      "project_id": 3,
    },
    Object {
      "coverage": "100.00",
      "project_locations.0": 3,
    },
    Object {
      "coverage": "200.00",
      "results": 6,
    },
    Object {
      "coverage": "100.00",
      "status": 3,
    },
    Object {
      "coverage": "100.00",
      "success_story": 3,
    },
    Object {
      "coverage": "633.33",
      "third_parties.0": 19,
    },
    Object {
      "coverage": "100.00",
      "title": 3,
    },
  ],
}
`;

exports[`Helper getCoverageReport "@eubfr/ingestion-quality-analyzer" provides reports for number of occurences of a given field and its coverage Works with results of wifi4eu/xls 1`] = `
Object {
  "meta": Object {
    "computed_key": "foo",
    "created_by": "bar",
    "last_modified": "baz",
  },
  "report": Array [
    Object {
      "call_year": 3,
      "coverage": "100.00",
    },
    Object {
      "coverage": "100.00",
      "description": 3,
    },
    Object {
      "coverage": "200.00",
      "third_parties.0": 6,
    },
    Object {
      "coverage": "100.00",
      "project_id": 3,
    },
    Object {
      "coverage": "400.00",
      "project_locations.0": 12,
    },
    Object {
      "coverage": "100.00",
      "project_website": 3,
    },
    Object {
      "coverage": "100.00",
      "timeframe": 3,
    },
    Object {
      "coverage": "100.00",
      "title": 3,
    },
  ],
}
`;
